[CORL-2381] Persist seen comments in the back-end

nick-funk opened the pull request:
## What does this PR do?

Persists seen comment state in the back-end so it can be a shared state between devices/logins for the same user.

## What changes to the GraphQL/Database Schema does this PR introduce?

- Adds a `seen` flag to the `Comment` type
- Adds a `markCommentAsSeen` mutation to the `Mutators` list

## How do I test this PR?

- Create the indexes listed in the deploy notes below
- Visit the stream as a logged in user and see that you can key through or refresh the stream and have comments go yellow to white
- Load up on your phone, another tab/browser and sign in as the same user, see that seen state is persisted to your user account
- Check your mongo for records in the `seenComments` collection
- set the following index, the `30` is 30 seconds (source: https://docs.mongodb.com/manual/core/index-ttl/)
    ```
    db.seenComments.createIndex({ lastSeenAt: 1 }, { expireAfterSeconds: 30 });
    ```
- wait 30 seconds
- watch your seenComments records get cleaned up after they're 30 seconds old
 
## How do we deploy this PR?

Create the following indexes in Mongo:

```
db.seenComments.createIndex({ tenantID: 1, storyID: 1, userID: 1 });
db.seenComments.createIndex({ lastSeenAt: 1 }, { expireAfterSeconds: 604800 });
```

Where `604800` is `7 days * 24 hours * 60 minutes * 60 seconds` for a TTL in seconds of how long to keep the seenComments records before deleting them gracefully within Mongo.

nick-funk committed:
Create initial implementation of seenComments storage and retrieval

Includes caching through a DataLoader to save the database
from loading the same large document numerous times.

Currently saves the seen comments as they're loaded in stream.

nick-funk committed:
Create preliminary automated job to clean up old seenComments records

nick-funk committed:
Make seen comments cron jobs configurable via env vars

nick-funk committed:
Create preliminary back-end for marking single comments as seen

nick-funk committed:
Remove await from executeBulkOperations

nick-funk committed:
Create mark as seen mutation that triggers on comment focus

nick-funk committed:
Remove client side seen marking from CommentContainer

nick-funk committed:
Update snapshots for back end `not-seen` tagging of comments

nick-funk committed:
Update tests and optimistic responses to check for seen state

nick-funk committed:
Properly implement the `seenComments` clean up bulk operation

kabeaty commented on src/core/server/models/comment/comment.ts:
```diff
@@ -143,6 +143,8 @@ export interface Comment extends TenantResource {
    */
   deletedAt?: Date;
+
+  seen?: boolean;
```
Can you add a comment here for `seen`?

nick-funk replied:
Yup! Added!

kabeaty commented on src/core/server/models/seenComments/seenComments.ts:
```diff
@@ -0,0 +1,126 @@
+
+import { TenantResource } from "../tenant";
+
+export interface SeenComments extends TenantResource {
```
What do you think about adding some comments here for `lastSeenAt`, `isDeleting`, `comments` etc.?

nick-funk replied:
Comments have been added!

nick-funk committed:
Fix issue where seen comments batching weren't working

wyattjoh commented on src/core/server/cron/seenCommentsCleanup.ts:
```diff
@@ -0,0 +1,74 @@
+  return { name: NAME, schedulers: [job] };
+}
+
+const cleanupSeenComments: ScheduledJobCommand<Options> = async ({
```
Why not instead utilize a TTL index in MongoDB?

https://docs.mongodb.com/manual/core/index-ttl/

It's designed to do this automatically.

nick-funk committed:
Fix default CRON interval for seenComments cleanup

nick-funk committed:
Show all comments as seen for logged out users

Also make the focus traversal show for all users whether
they're logged in our out. No reason they can't have a nice
blue border when using the C key even if they're logged out.

nick-funk committed:
Remove seenCommentsCleanup task in favour of a TTL index

`db.seenComments.createIndex({ lastSeenAt: 1 }, { expireAfterSeconds: 604800 });`

Where `604800` is 7 days * 24 hours * 60 minutes * 60 seconds for a
TTL in seconds of how long to keep the seenComments records.

nick-funk committed:
Update snapshots for to match the new server-side seenComments

kabeaty commented on src/core/server/models/seenComments/seenComments.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+}
+
+export async function retrieveSeenCommentsForDeletion(
```
Can this be removed now?

nick-funk replied:
Yup! Removed it!

kabeaty commented on src/core/client/stream/tabs/Comments/Comment/CommentContainer.tsx:
```diff
@@ -419,7 +432,7 @@ export const CommentContainer: FunctionComponent<Props> = ({
       // Added for keyboard shortcut support.
       data-key-stop
-      data-not-seen={seen || !commentSeenEnabled ? undefined : true}
+      data-not-seen={commentSeenEnabled ? !comment.seen : false}
```
I think this needs to be something like `commentSeenEnabled && comment.seen ? undefined : true`. 

In `KeyboardShortcuts`, in `toKeyStop`, it's just checking whether `data-not-seen` is there or not. So, setting it to `false` here is making it pass the check as not seen, and the `Z` key shortcut therefore isn't working as expected.

nick-funk replied:
I inverted it to be `commentSeenEnabled && !comment.seen ? true : undefined` to make it more in-line with the fact the tag is `data-not-seen`. Other than that, updated and good catch!

kabeaty reviewed (changes requested):
Looks good overall! I left a few comments. Also, it looks like the `CommentContainer` styles for the logged-out state are missing some padding now. Can you take a look at this?

nick-funk committed:
Add comments to seenComments models and retrieval/update operations

nick-funk committed:
Properly tage `comment-not-seen` as it was with client side seen comments

If the comment is not seen, tag it, otherwise, don't tag it.

nick-funk committed:
Merge remote-tracking branch 'origin/develop' into feat/CORL-2381-seenComments

nick-funk commented:
> Looks good overall! I left a few comments. Also, it looks like the `CommentContainer` styles for the logged-out state are missing some padding now. Can you take a look at this?

I have no idea what's going on with the padding, but it's happening on develop. Logging a bug ticket.

nick-funk committed:
Don't make comment vertical padding dependent on commentSeen being enabled

nick-funk committed:
Update setTraversalFocus hook for server side commentSeen

- Moved event emitter logic to the markSeenMutation
- Separate commentSeenEnabled and canCommitCommentSeen checks
to selectively enable certain features

nick-funk committed:
Update snapshots

nick-funk commented:
> > Looks good overall! I left a few comments. Also, it looks like the `CommentContainer` styles for the logged-out state are missing some padding now. Can you take a look at this?
> 
> I have no idea what's going on with the padding, but it's happening on develop. Logging a bug ticket.

Found it, also found more bugs in traversal. All should be fixed up now. I noticed it seems to not want to let Z key do anything unless signed in. We might want to revisit this behaviour, but at least this no longer breaks existing padding/traversal flow now!

kabeaty reviewed (approved):
Looks great! Thanks for making the updates.

kodiakhq[bot] committed:
Merge branch 'develop' into feat/CORL-2381-seenComments
